pandas/io/stata.py

```diff
@@ -1411,150 +1407,155 @@ def read(self, nrows=None, convert_dates=None,
              convert_categoricals=None, index=None,
              convert_missing=None, preserve_dtypes=None,
              columns=None, order_categoricals=None):
-
         # Handle empty file or chunk.  If reading incrementally raise
         # StopIteration.  If reading the whole thing return an empty
         # data frame.
         if (self.nobs == 0) and (nrows is None):
             self._can_read_value_labels = True
             self._data_read = True
+            self.close()
             return DataFrame(columns=self.varlist)
 
-        # Handle options
-        if convert_dates is None:
-            convert_dates = self._convert_dates
-        if convert_categoricals is None:
-            convert_categoricals = self._convert_categoricals
-        if convert_missing is None:
-            convert_missing = self._convert_missing
-        if preserve_dtypes is None:
-            preserve_dtypes = self._preserve_dtypes
-        if columns is None:
-            columns = self._columns
-        if order_categoricals is None:
-            order_categoricals = self._order_categoricals
-
-        if nrows is None:
-            nrows = self.nobs
-
-        if (self.format_version >= 117) and (self._dtype is None):
-            self._can_read_value_labels = True
-            self._read_strls()
-
-        # Setup the dtype.
-        if self._dtype is None:
-            dtype = []  # Convert struct data types to numpy data type
-            for i, typ in enumerate(self.typlist):
-                if typ in self.NUMPY_TYPE_MAP:
-                    dtype.append(('s' + str(i), self.byteorder +
-                                  self.NUMPY_TYPE_MAP[typ]))
-                else:
-                    dtype.append(('s' + str(i), 'S' + str(typ)))
-            dtype = np.dtype(dtype)
-            self._dtype = dtype
-
-        # Read data
-        dtype = self._dtype
-        max_read_len = (self.nobs - self._lines_read) * dtype.itemsize
-        read_len = nrows * dtype.itemsize
-        read_len = min(read_len, max_read_len)
-        if read_len <= 0:
-            # Iterator has finished, should never be here unless
-            # we are reading the file incrementally
+        try:
```

Comment:
you don't nee dthe try/except over everthing here. pls narrow down as much as possible.